File: dsa-problems/LinkedLists/solutions/python/LNK-013-shuttle-ticket-rotate-blocks.py
import sys

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

def rotate_list(head: ListNode, length: int, k: int):
    if length <= 1 or k % length == 0:
        tail = head
        while tail.next:
            tail = tail.next
        return head, tail
    
    k = k % length
    moves = length - k
    
    new_tail = head
    for _ in range(moves - 1):
        new_tail = new_tail.next
        
    new_head = new_tail.next
    new_tail.next = None
    
    temp = new_head
    while temp.next:
        temp = temp.next
    temp.next = head
    
    return new_head, new_tail

def rotate_blocks(head: ListNode, b: int, k: int) -> ListNode:
    if not head or b <= 0:
        return head
        
    dummy = ListNode(0)
    prev_tail = dummy
    curr = head
    
    while curr:
        block_head = curr
        block_tail = curr
        length = 1
        
        # Find block end
        for _ in range(b - 1):
            if not block_tail.next:
                break
            block_tail = block_tail.next
            length += 1
            
        next_block_head = block_tail.next
        block_tail.next = None # Detach
        
        # Rotate
        new_head, new_tail = rotate_list(block_head, length, k)
        
        # Attach
        prev_tail.next = new_head
        prev_tail = new_tail
        
        curr = next_block_head
        
    return dummy.next

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        dummy = ListNode()
        cur = dummy
        for _ in range(n):
            cur.next = ListNode(int(next(iterator)))
            cur = cur.next
            
        b = int(next(iterator))
        k = int(next(iterator))
        
        head = rotate_blocks(dummy.next, b, k)
        
        out = []
        while head:
            out.append(str(head.val))
            head = head.next
        print(" ".join(out))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/LinkedLists/solutions/javascript/LNK-013-shuttle-ticket-rotate-blocks.js
const readline = require("readline");

class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

function rotateList(head, len, k) {
  if (len <= 1 || k % len === 0) {
    let tail = head;
    while (tail.next) tail = tail.next;
    return [head, tail];
  }

  k = k % len;
  let moves = len - k;

  let newTail = head;
  for (let i = 0; i < moves - 1; i++) {
    newTail = newTail.next;
  }

  let newHead = newTail.next;
  newTail.next = null;

  let temp = newHead;
  while (temp.next) temp = temp.next;
  temp.next = head;

  return [newHead, newTail];
}

function rotateBlocks(head, b, k) {
  if (!head || b <= 0) return head;

  const dummy = new ListNode(0);
  let prevTail = dummy;
  let curr = head;

  while (curr) {
    let blockHead = curr;
    let blockTail = curr;
    let len = 1;

    for (let i = 0; i < b - 1 && blockTail.next; i++) {
      blockTail = blockTail.next;
      len++;
    }

    let nextBlockHead = blockTail.next;
    blockTail.next = null;

    const [newHead, newTail] = rotateList(blockHead, len, k);

    prevTail.next = newHead;
    prevTail = newTail;

    curr = nextBlockHead;
  }

  return dummy.next;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/).filter(x => x)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  
  const dummy = new ListNode(0);
  let cur = dummy;
  for (let i = 0; i < n; i++) {
    cur.next = new ListNode(parseInt(data[idx++], 10));
    cur = cur.next;
  }
  
  if (idx < data.length) {
      const b = parseInt(data[idx++], 10);
      const k = parseInt(data[idx++], 10);

      let head = rotateBlocks(dummy.next, b, k);
      const out = [];
      while (head) {
        out.push(head.val);
        head = head.next;
      }
      console.log(out.join(" "));
  }
});
